README.md
# inquisitive

<!-- badges: start -->
<!-- badges: end -->

A collection of exploratory data analysis serving two goals:

1) Clearly characterize source data issues to drive operational improvements
2) Gain insights into the structure and limits of the data to improve feature engineering and modeling

## Installation

You can install the development version of inquisitive from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("johnaclouse/inquisitive")
```

## Work flow

* Capture the starting point
  + The original tibble is used to calculate the impact of the suspected data
  + Do not change the original tibble until finished identifying suspect rows and columns
* Add suspect rows using [add_suspected_rows()](man/add_suspected_columns.Rd)
* Add suspect columns using add_suspected_columns()
* Plot the results using plot_suspect_data()
* Adjusticate the original tibble
  + Apply the function to the original tibble to remove all of the suspect data using 
  





NEWS.md
# inquisitive 0.1.0

* Plot_missingness_by_var now aggregates data if more than 5,000 rows are present and will show percentage of missing data in the plot.
* Updated previous fix and returned to using geom_raster for plot_missingness_by_var

# inquisitive 0.0.5

* Fixed bug where geom_raster is unable to handle more than 32,667 rows on some systems.

# inquisitive 0.0.4

* Moved the reason label from the body of the columns removed graph to a secondary x-asis to improve visibility of missing and present in the stacked chart.

# inquisitive 0.0.3

* Updateds to adjudicate_suspect_data and plot_suspect_data documentation

# inquisitive 0.0.2

* Bug fixes and appeasements to R CMD check

# inquisitive 0.0.1

* Added plot_suspect_data functions included adjudication helper function.

# inquisitive 0.0.0.9001

* Beginning consolidation of EDA tools from multiple sources


R/add_suspected_data.R
#' Constructor function for suspected data tibble structure
#'
#' Suspected data identified by row or column is stored in a common structure
#' supporting independent collection of suspicious data prior to final adjudication.
#'
#' @return A tibble structured to support suspecting data
#' @export
#'
#' @examples
#' suspect_data_constructor()
suspect_data_constructor <- function() {
  tibble(
    type = factor(character(), levels = c("row", "column")),
    reason = character(),
    data =  list())
}


#' Identify rows of a data set to be marked as suspicious
#'
#' Given a common tibble of interest, a subset of the data can be identified as
#' suspicious and added to the suspect_data structure for analysis and
#' visualization.
#'
#' @param x a subset of rows from the tibble of interest
#' @param reason a character string describing why this particular subset is
#'   suspect
#' @param suspect_data a specifically structure tibble
#'
#' If suspect_data is NULL, a new suspect_data tibble will be constructed and
#' the identified data added to that structure.
#'
#' @return a tibble with the suspect data structure
#' @export
#'
#' @examples
#' df <- create_suspicious_data()
#' suspect_data <-
#'   df %>%
#'   filter(cut == "Fair") %>%
#'   add_suspected_rows(reason = "Cut was not good or better")
#'
# suspect_data <-
#   df %>%
#   filter(price < 5000) %>%
#   add_suspected_rows(reason = "Price was less than 5000", suspect_data)
#
# suspect_data <-
#   df %>%
#   filter(table < 57) %>%
#   add_suspected_rows(reason = "Table was less than 57", suspect_data)
add_suspected_rows <- function(x,
                              reason,
                              suspect_data = NULL){

  if (is.null(suspect_data)) suspect_data <- suspect_data_constructor()

  suspect_data %>%
    add_row(tibble::tibble_row(
      type = "row",
      reason = reason,
      data = list(x)))
}

#' Identify column(s) to be marked as suspicious
#'
#' Given a common tibble of interest, a subset of the data can be identified as
#' suspicious and added to the suspect_data structure for analysis and
#' visualization.
#'
#' @param x a character string containin the names of suspect columns from the tibble of interest
#' @param reason a character string describing why this particular subset is
#'   suspect
#' @param suspect_data a specifically structure tibble
#'
#' If suspect_data is NULL, a new suspect_data tibble will be constructed and
#' the identified data added to that structure.
#'
#' @return a tibble with the suspect data structure
#' @export
#'
#' @examples
#' df <- create_suspicious_data()
#' suspect_data <-
#' df %>%
#'   select(clarity, y) %>%
#'   add_suspected_columns(reason = "These columns are hinky")
#'
#' suspect_data <-
#'   df %>%
#'   select(depth) %>%
#'   add_suspected_columns(reason = "This column is not deep enough", suspect_data)
add_suspected_columns <- function(x,
                                 reason,
                                 suspect_data = NULL){

  if (is.null(suspect_data)) suspect_data <- suspect_data_constructor()

  suspect_data %>%
    add_row(tibble::tibble_row(
      type = "column",
      reason = reason,
      data = list(x)))
}


R/create_suspicious_data.R
#' Create sample data for the inquisitive
#'
#' A function to create a set of data useful for testing and evaluating inquisitive tools.
#'
#' @param seed an integer used to set the random number generator state.
#' @return A tibble.
#' @export
#'
#' @examples
#' create_suspicious_data()
create_suspicious_data <- function(seed = set.seed(23)){
  . <- key_date <- NULL # satisfy R CMD check issue with NSE
  set.seed(seed)

  ds <- ggplot2::diamonds %>%
    mutate(id = row_number(),
           key_date = as.Date(as.character(20190101), format = "%Y%m%d") +
             round(stats::runif(nrow(.)) * 5 * 365)) %>%
    relocate(id, key_date)

  ds[1:20, "cut"] <- NA
  ds[40:100, "price"] <- NA
  ds[10:60, "table"] <- NA
  ds$color <- ifelse(stats::runif(nrow(ds)) > 0.33, NA, ds$color)
  ds$x <- ifelse(stats::runif(nrow(ds)) > 0.25, NA, ds$x)
  ds$y <- ifelse(stats::runif(nrow(ds)) > 0.5, NA, ds$y)
  ds$z <- ifelse(stats::runif(nrow(ds)) > 0.95, NA, ds$z)
  ds$depth[c(1:20000, 25000:30000)] <- NA
  ds[ds$key_date < "2020-01-01", "depth"] <- NA
  ds[ds$key_date > "2022-01-01", "clarity"] <- NA
  ds
}



R/adjudicate_suspect_data.R
#' Adjudicate suspected data removing it from the original data set
#'
#' Per the contents of the suspect_data, remove the rows and columns identified.
#'
#' @param ds tibble containing the original data set
#' @param suspect_data a suspect_date object to be applied to the original data
#'   set
#'
#' @return an adjudicated_data object containing the original and adjudicated
#'   data sets and associated metadata.
#' @export
#'
#' @examples
#' original_data <- create_suspicious_data()
#' suspect_data <-
#'   original_data %>%
#'   filter(cut == "Fair") %>%
#'   add_suspected_rows(reason = "Cut was not good or better")
#'
#' suspect_data <-
#'   original_data %>%
#'   filter(price < 5000) %>%
#'   add_suspected_rows(reason = "Price was less than 5000", suspect_data)
#'
#' suspect_data <-
#'   original_data %>% filter(table < 57) %>%
#'   add_suspected_rows(reason = "Table was less than 57", suspect_data)
#'
#' suspect_data <-
#'   original_data %>%
#'   select(clarity, y) %>%
#'   add_suspected_columns(reason = "These columns are hinky", suspect_data)
#'
#' suspect_data <-
#'   original_data %>%
#'   select(depth) %>%
#'   add_suspected_columns(reason = "This column is not deep enough", suspect_data)
#'
#' adjudicated_data <- adjudicate_suspect_data(original_data, suspect_data)
adjudicate_suspect_data <- function(ds, suspect_data){

  type <- data <- NULL

  original_data <- ds
  original_dim <- dim(ds)
  original_na_count <- sum(is.na(ds))
  original_present_count <- sum(!is.na(ds))

  excluded_rows <-
    suspect_data %>%
    filter(type == "row") %>%
    select(data) %>%
    tidyr::unnest(data) %>%
    distinct()

  # remove excluded rows
  ds <- anti_join(
    ds,
    excluded_rows,
    by = names(excluded_rows)
  )

  excluded_columns <-
    suspect_data %>%
    filter(type == "column") %>%
    select(data) %>%
    tidyr::unnest(data) %>%
    names()

  # remove excluded columns
  ds <-
    ds %>%
    select(-any_of(excluded_columns))

  adjudicated_dim = dim(ds)

  list(
    original_data = original_data,
    adjudicated_data = ds,
    original_dim = original_dim,
    original_na_count = original_na_count,
    original_present_count = original_present_count,
    adjudicated_dim = adjudicated_dim,
    adjudicated_na_count = sum(is.na(ds)),
    adjudicated_present_count = sum(!is.na(ds)),
    adjudicated_pct = adjudicated_dim / original_dim,
    adjudicated_diff = original_dim - adjudicated_dim
  )
}


R/plot_missingness_by_var.R
#' Plot missingness by variable
#'
#' Build a ggplot depicting missingness in a data frame. The data is first
#' turned into a lithographic copy of the frame with only 'missing' or 'present'
#' by \code{\link{as_lith_tbl}}.
#'
#' If the data set contains more than 5,000 rows, the data will be aggregated
#' and the percentage of missing data per aggregated element will be reported.
#'
#' If a reference column is specified, the data frame will be sorted by the
#' reference column prior to plotting
#'
#' @param ds a data frame to be plotted
#' @param reference_column an unquoted column name
#' @param number_of_axis_lables an integer specifying maximum number of labels
#'   for the x axis
#' @param date_format_fun a function specifying the formatting of of the x axis
#'   when the column selected is of type Date.
#'
#' @return ggplot object
#' @export
#'
#' @examples
#' ds <- inquisitive::create_suspicious_data()
#' plot_missingness_by_var(ds, reference_column = key_date)
#' plot_missingness_by_var(ds, reference_column = key_date, date_format_fun =
#'  function(x) {format(x, "%b- %Y")})
#' plot_missingness_by_var(ds, reference_column = carat)
#' plot_missingness_by_var(ds)
plot_missingness_by_var <- function(ds,
                                    reference_column = NULL,
                                    number_of_axis_lables = 25,
                                    date_format_fun = function(x) {format(x, "%Y-%m")}) {

  . <- column <- row_identifier <- name <- value <- NULL

  number_of_axis_lables <- pmin(number_of_axis_lables,
                                nrow(ds))

  column_order <-
    purrr::map_df(ds, ~ sum(is.na(.))) %>%
    tidyr::pivot_longer(cols = everything(),
                        names_to = "column",
                        values_to = "count") %>%
    arrange(count) %>%
    pull(column)

  ds <- as_lith_tbl(ds, {
    {
      reference_column
    }
  })

  subject_label <- (rlang::enexpr(reference_column))

  if (is.null(subject_label)) {
    subject_label <- "row identifier"
  } else {
    ds <- ds %>%
      select(-{{subject_label}})
  }

  ds <-
    ds %>%
    mutate(across(tidyselect::where(is.factor), ~ if_else(. == "present", 0, 1)))

  if (nrow(ds) > 5000) {
    observations_per_element <-  ceiling(nrow(ds) / 5000)
    ds <-
      ds %>%
      arrange(reference_column) %>%
      group_by(row_identifier = row_number() %/% observations_per_element) %>%
      summarize(across(everything(), mean))

    caption <- glue::glue("Darker color indicates greater percentage of missing data\nEach bar represents {observations_per_element} observations")

  } else {
    ds <- ds %>%
      arrange(reference_column) %>%
      mutate(row_identifier = row_number())

    caption <- "Dark color indicates missing data"
  }

  x_axis_labels <-
    ds %>%
    # row_identifier is required here to tie labels to original data
    distinct(row_identifier,
             reference_column) %>%
    arrange(reference_column, row_identifier) %>%
    slice(seq(1, nrow(.), by = floor(nrow(.) / number_of_axis_lables)))

  plot_data <-
    ds %>%
    tidyr::pivot_longer(cols = -c(row_identifier,
                                  reference_column)) %>%
    mutate(name = factor(name, levels = column_order))

  ggplot(data = plot_data,
         aes(
           x = row_identifier,
           y = name,
           fill = value
         )) +
    geom_raster() +
    scale_x_continuous(
      breaks = x_axis_labels$row_identifier,
      labels = x_axis_labels$reference_column
    ) +

    scale_fill_viridis_c(
      option = "E",
      direction = -1,
      limits = c(0,1)
    ) +
    ggplot2::labs(
      title = paste("Missing data in features by", quo_name(subject_label)),
      caption = caption,
      x = subject_label,
      y = ""
    ) +
    ggplot2::theme_minimal() +
    ggplot2::theme(
      axis.text.x = ggplot2::element_text(
        angle = 45,
        vjust = 1,
        hjust = 1
      ),
      legend.position = "none",
      panel.grid.minor.x = element_blank()
    )
}







#' Create a copy of a tibble replacing all of the elements with a binary missing
#' or present factor.
#'
#' The function takes a tibble composed of arbitrary data types and creates a
#' copy of the structure but replaces all of the values with either 'present' or
#' 'missing'. Something like a photographic lithograph.
#'
#' In addition to replace the values, two columns are added: row_identifier and
#' reference_column. These are used to allow for the transformed data to be
#' incorporated into subsequent analysis and visualization. If not reference
#' column is provided, the row_identifier column will be used as the reference_column.
#'
#' @param ds a tibble to be developed as a data lithograph
#' @param reference_column an unquoted character string designating a reference
#'   column to be preserved for sorting and future visualization. If no column
#'   is designated, the function will use the row_identifier column.
#'
#' @return a tibble comprised of the original columns containing present/missing
#'   values and the reference column in its original state.
#' @export
#'
#' @examples
#' ds <- create_suspicious_data()
#' as_lith_tbl(ds, reference_column = carat) %>% head()
#' as_lith_tbl(ds, reference_column = key_date) %>% head()
#' as_lith_tbl(ds) %>% head()
as_lith_tbl <- function(ds, reference_column = NULL){

  reference_column <- rlang::enquo(reference_column)


  if (rlang::quo_is_null(reference_column)) {
    ds$reference_column <- seq(1, nrow(ds))
  } else if (rlang::quo_is_symbol(reference_column)) {
    ds$reference_column <- rlang::eval_tidy(ds[[rlang::ensym(reference_column)]])
  } else {
    stop(paste(
      "Expected a symbol for reference_column but found",
      class(rlang::get_expr(reference_column))
    ))
  }

  # Base R solution:
  # if (is.null(substitute(reference_column))) {
  #   ds$reference_column <- ds$row_identifier
  # } else {
  #   ds$reference_column <- ds[[deparse(substitute(reference_column))]]
  # }

  ds %>%
    mutate(across(
      c(everything(), -reference_column),
      ~ factor(if_else(is.na(.), "missing", "present"))
    ))
}



R/plot_suspect_data.R
plot_suspect_rows <- function(suspect_data) {
  type <- data <- row_count <- missing_elements <- present_elements <- NULL
  reason <- present <- rows <- missingness <-

    axis_expansion <- 2

  suspect_row_plot_data <-
    suspect_data %>%
    filter(type == "row") %>%
    rowwise() %>%
    mutate(
      row_count = as.numeric(nrow(data)),
      missing_elements = as.numeric(sum(is.na(data))),
      present_elements = as.numeric(sum(!is.na(data))),
      missing = row_count * missing_elements / (missing_elements + present_elements),
      present = row_count * present_elements / (missing_elements + present_elements)
    ) %>%
    ungroup() %>%
    select(-data) %>%
    mutate(reason = forcats::fct_reorder(reason, row_count, .desc = F)) %>%
    arrange(desc(row_count))

  suspect_row_plot_data %>%
    tidyr::pivot_longer(cols = c(missing, present),
                        names_to = "missingness",
                        values_to = "rows") %>%
    ggplot() +
    geom_col(aes(x = reason, y = rows, fill = missingness)) +
    scale_x_discrete(
      expand = expansion(add = c(axis_expansion, 0)),
      position = "top"
    ) +
    ggplot2::scale_fill_manual(name = "Data elements",
                               values = c("missing" = "#FFEA46",
                                          "present" =  "#575C6D")
    ) +
    labs(
      x = "Rows removed",
      y = "Count of rows removed") +
    coord_flip() +
    theme_minimal() +
    theme(aspect.ratio = 1,
          legend.position = "none",
          axis.title.x.bottom = element_text(margin = margin(t = 10, r = 0, b = 0, l = 0)),
          axis.title.y.right = element_text(margin = margin(t = 0, r = 0, b = 0, l = 10))

          # legend.position = "none",
          # axis.title.y = element_blank()
    )
}
# plot_suspect_rows(suspect_data)

plot_suspect_columns <- function(suspect_data) {
  type <- data <- column_name <- row_count <- missing_elements <- NULL
  present_elements <- present <- rows <- missingness <- reason <- NULL


  axis_expansion <- 2

  suspect_column_plot_data <-
    suspect_data %>%
    filter(type == "column") %>%
    select(-type) %>%
    # group_by(reason) %>%
    mutate(
      column_name = purrr::map(data, ~ names(.)),
    ) %>%
    tidyr::unnest(column_name) %>%
    mutate(
      row_count = purrr::map2_dbl(data, column_name, ~ nrow(.x[.y])),
      missing_elements = purrr::map2_dbl(data, column_name, ~ sum(is.na(.x[.y]))),
      present_elements = purrr::map2_dbl(data, column_name, ~ sum(!is.na(.x[.y]))),
      missing = row_count * missing_elements / (missing_elements + present_elements),
      present = row_count * present_elements / (missing_elements + present_elements)
    ) %>%
    ungroup() %>%
    select(-data) %>%
    mutate(column_name = forcats::fct_reorder(column_name, present, .desc = FALSE)) %>%
    arrange(desc(present))

  suspect_column_plot_data %>%
    tidyr::pivot_longer(cols = c(missing, present),
                        names_to = "missingness",
                        values_to = "rows") %>%
    ggplot() +
    geom_col(aes(x = as.numeric(column_name), y = rows, fill = missingness)) +
    # ggtext::geom_richtext(aes(x = column_name, y = 1, label = reason),
    #                       angle = 90,
    #                       hjust = 0,
    #                       vjust = 0.5,
    #                       nudge_y = 0.01 * max(suspect_column_plot_data$row_count, na.rm = TRUE),
    #                       size = 3,
    #                       color = "black",
    #                       fill = "white") +


    # scale_x_discrete(expand = expansion(add = c(axis_expansion, 0)),
  #                  position = "top"
  #                  ) +

  scale_x_continuous(expand = expansion(add = c(axis_expansion, 0)),
                     position = "top",
                     breaks = unique(as.numeric(suspect_column_plot_data$column_name)),
                     labels = unique(suspect_column_plot_data$column_name),
                     sec.axis = dup_axis(
                       labels = suspect_column_plot_data$reason)
  ) +

    ggplot2::scale_fill_manual(
      values = c("missing" = "#FFEA46",
                 "present" =  "#575C6D")
    ) +

    labs(
      x = "Columns removed",
      y = "Count of data elements removed") +
    theme_minimal() +
    theme(aspect.ratio = 1,
          legend.position = "none",
          axis.text.x.top = element_text(angle = 45, hjust = 0, vjust = 0),
          axis.text.x.bottom = element_text(angle = 90, vjust = 0.5)
    )
}
# plot_suspect_columns(suspect_data)


plot_suspect_missingness <- function(adjudicated_ds) {

  dataset <- elements <- missingness <- NULL

  synopsis <- list()

  synopsis$original$missing <- adjudicated_ds$original_na_count
  synopsis$original$present <- adjudicated_ds$original_present_count

  synopsis$adjudicated$missing <- adjudicated_ds$adjudicated_na_count
  synopsis$adjudicated$present <- adjudicated_ds$adjudicated_present_count

  missingness_plot_data <-
    bind_rows(synopsis, .id = "dataset") %>%
    tidyr::pivot_longer(cols = -dataset,
                        names_to = "missingness",
                        values_to = "elements")


  missingness_plot_data %>%
    ggplot() +
    geom_col(aes(x = dataset, y = elements, fill = missingness)) +
    # geom_text(aes(x = column_name, y = 1, label = reason),
    #           angle = 90,
    #           hjust = 0,
    #           vjust = 0.5,
    #           nudge_y = 0.01 * max(suspect_column_plot_data$row_count, na.rm = TRUE),
    #           size = 3,
    #           color = "white",) +
    scale_x_discrete(# expand = expansion(add = c(0, 0)),
      position = "top") +
    scale_y_continuous(position = "right", labels = scales::comma) +
    ggplot2::scale_fill_manual(
      values = c("missing" = "#FFEA46",
                 "present" =  "#575C6D")
    ) +

    labs(
      x = "Missingness by pipeline stage",
      y = "Count of data elements") +
    theme_minimal() +
    theme(aspect.ratio = 1,
          legend.position = c(1.4,0.13),

          plot.margin = margin(t = 10,
                               r = 10,
                               b = 10,
                               l = 10),

          # legend.box.spacing = unit(0, "points"),
          # legend.background = element_blank(),
          # legend.spacing.x = unit(1.0, 'cm'),
          # legend.margin = margin(3, .5, .5, .5, "lines"),
          # legend.key.height = unit(1, "lines")
          axis.text.x = element_text(angle = 60, hjust = 0),
          axis.title.x.top = element_text(size = rel(0.8), margin = margin(t = 0, r = 0, b = 10, l = 0)),
          axis.title.y.right = element_text(size = rel(0.8), margin = margin(t = 0, r = 0, b = 0, l = 10))
          # axis.text.x = element_text(angle = 60, hjust = 0.)
          # axis.title.y = element_blank()
    )

}
# plot_suspect_missingness(adjudicated_data)




plot_suspect_area <- function(adjudicated_ds){

  # adjudication_pct <- adjudicated_ds$adjudicated_dim / adjudicated_ds$original_dim
  # adjudication_diff <- adjudicated_ds$original_dim - adjudicated_ds$adjudicated_dim
  adjudication_pct <- adjudicated_ds$adjudicated_pct
  adjudication_diff <- adjudicated_ds$adjudicated_diff

  rejected_data_pct <- round(100 * (1 - prod(adjudication_pct)),1)
  rejected_rows <- format(adjudicated_ds$adjudicated_diff[1], big.mark = ",", scientific = FALSE)
  rejected_columns <- format(adjudicated_ds$adjudicated_diff[2], big.mark = ",", scientific = FALSE)
  retained_data_pct <- round(100 * (prod(adjudication_pct)),1)
  retained_rows <- format(adjudicated_ds$adjudicated_dim[1], big.mark = ",", scientific = FALSE)
  retained_columns <- format(adjudicated_ds$adjudicated_dim[2], big.mark = ",", scientific = FALSE)

  ggplot() +
    # background
    geom_rect(
      aes(xmin = 0, xmax = 1,
          ymin = 0, ymax = 1),
      color = "#B22222",
      fill = "#Fbebeb",
      linewidth = 1,
    ) +
    # rows and columns by percentage
    geom_rect(
      aes(xmin = 0, xmax = adjudication_pct[2],
          ymin = 0, ymax = adjudication_pct[1]),
      color = "black",
      fill = "gray92",
      linewidth = 1
    ) +
    # arrow from 100% of original data to scaled processed data representation
    geom_segment(
      aes(
        x = 1 - (1 - adjudication_pct[2]) * 0.02,
        y = 1 - (1 - adjudication_pct[1]) * 0.02,
        xend = adjudication_pct[2] + (1 - adjudication_pct[2]) * 0.02,
        yend = adjudication_pct[1] + (1 - adjudication_pct[1]) * 0.02,
      ),
      color = "#B22222",
      arrow = arrow(length = unit(0.5, "cm"))
    ) +
    annotate(
      "text",
      label = glue::glue(
        "Rejected data: {rejected_data_pct}%\n",
        "Rows: {rejected_rows}\n",
        "Columns: {rejected_columns}"
      ),
      x = 0.05,
      y = mean(c(adjudication_pct[1], 1)),
      lineheight = 0.9,
      hjust = 0,
      color = "#B22222"
    ) +
    annotate(
      "text",
      label = glue::glue(
        "Retained data: {retained_data_pct}%\n",
        "Rows: {retained_rows}\n",
        "Columns: {retained_columns}"
      ),
      x = 0.05,
      y = adjudication_pct[1] / 2,
      lineheight = 0.9,
      hjust = 0,
      color = "black"
    ) +
    scale_x_continuous(expand = expansion(mult = 0)) +
    scale_y_continuous(expand = expansion(mult = 0)) +
    theme_minimal() +
    theme(aspect.ratio = 1,
          legend.position = "none",
          axis.title = element_blank(),
          plot.margin = unit(c(0, 0, 0, 0), "cm")
    )
}
# plot_suspect_area(adjudicated_data)

#' Plot the effects of removing suspected data against the original data set
#'
#' @param suspect_ds a suspect_data object
#' @param adjudicated_ds an adjudicated data object
#'
#' @return ggplot object
#' @export
#'
#' @examples
#' df <- create_suspicious_data()
#' suspect_data <-
#'   df %>%
#'   filter(cut == "Fair") %>%
#'   add_suspected_rows(reason = "Cut was not good or better")
#'
#' suspect_data <-
#'   df %>%
#'   filter(price < 5000) %>%
#'   add_suspected_rows(reason = "Price was less than 5000", suspect_data)
#'
#' suspect_data <-
#'   df %>%
#'   filter(table < 57) %>%
#'   add_suspected_rows(reason = "Table was less than 57", suspect_data)
#'
#' suspect_data <-
#'   df %>%
#'   select(clarity, y) %>%
#'   add_suspected_columns(reason = "These columns are hinky",
#'                         suspect_data)
#'
#' adjudicated_data <- adjudicate_suspect_data(df, suspect_data)
#'
#' plot_suspect_data(suspect_ds = suspect_data, adjudicated_ds = adjudicated_data)
plot_suspect_data <- function(suspect_ds, adjudicated_ds) {
  suspect_row_plot <- plot_suspect_rows(suspect_ds)
  suspect_missingness_plot <- plot_suspect_missingness(adjudicated_ds)
  suspect_area_plot <- plot_suspect_area(adjudicated_ds)
  suspect_columns_plot <- plot_suspect_columns(suspect_ds)

  # patchwork::wrap_plots(
  suspect_columns_plot +
    patchwork::plot_spacer() +
    suspect_area_plot +
    suspect_row_plot +
    patchwork::inset_element(suspect_missingness_plot,
                             left = 0.25,
                             bottom = 0,
                             right = 1.1,
                             top = 3.2)
  # patchwork::plot_layout(guides = "collect")
}
